diff --git i/options/posix/generic/unistd.cpp w/options/posix/generic/unistd.cpp
index 6ef827df..71c2bb42 100644
--- i/options/posix/generic/unistd.cpp
+++ w/options/posix/generic/unistd.cpp
@@ -1430,6 +1430,14 @@ int access(const char *path, int mode) {
 	return 0;
 }
 
+int euidaccess(const char *path, int mode) {
+	return faccessat(AT_FDCWD, path, mode, AT_EACCESS);
+}
+
+int eaccess(const char *path, int mode) {
+	return euidaccess(path, mode);
+}
+
 namespace {
 	FILE *user_shell_global_file; // Used by setusershell/getusershell/endusershell.
 
diff --git i/options/posix/include/unistd.h w/options/posix/include/unistd.h
index 93020c94..8647fb8a 100644
--- i/options/posix/include/unistd.h
+++ w/options/posix/include/unistd.h
@@ -477,6 +477,11 @@ char *get_current_dir_name(void);
 pid_t gettid(void);
 #endif
 
+#if defined(_GNU_SOURCE)
+int euidaccess(const char *__path, int __mode);
+int eaccess(const char *__path, int __mode);
+#endif
+
 int getentropy(void *__buffer, size_t __size);
 
 int pipe2(int *__pipefd, int __flags);
