diff --git i/options/ansi/include/string.h w/options/ansi/include/string.h
index 3980418c..97543cd4 100644
--- i/options/ansi/include/string.h
+++ w/options/ansi/include/string.h
@@ -60,11 +60,8 @@ size_t strlen(const char *__s);
 #ifndef __MLIBC_ABI_ONLY
 
 /* POSIX extensions. */
-#if defined(_GNU_SOURCE)
-char *strerror_r(int __errnum, char *__buffer, size_t __size) __asm__("__gnu_strerror_r");
-#else
 int strerror_r(int __errnum, char *__buffer, size_t __size);
-#endif
+char *gnu_strerror_r(int __errnum, char *__buffer, size_t __size) __asm__("__gnu_strerror_r");
 
 /* Handling the basename mess:
  * If <libgen.h> is included *at all*, we use the XPG-defined basename
