diff --git a/options/internal/include/mlibc/allocator.hpp b/options/internal/include/mlibc/allocator.hpp
index 4dfbb71d..560a94b7 100644
--- a/options/internal/include/mlibc/allocator.hpp
+++ b/options/internal/include/mlibc/allocator.hpp
@@ -10,6 +10,12 @@
 
 struct VirtualAllocator {
 public:
+#ifdef ORLIX_HOSTED_PAGE_SIZE
+	static constexpr size_t pagesize = ORLIX_HOSTED_PAGE_SIZE;
+#else
+	static constexpr size_t pagesize = 0x1000;
+#endif
+
 	uintptr_t map(size_t length);
 
 	void unmap(uintptr_t address, size_t length);
