diff --git i/options/posix/generic/grp.cpp w/options/posix/generic/grp.cpp
index 41f6d1d8..8c33fed2 100644
--- i/options/posix/generic/grp.cpp
+++ w/options/posix/generic/grp.cpp
@@ -130,7 +130,7 @@ namespace {
 			}
 		}
 
-		int err = ESRCH;
+		int err = ENOENT;
 		if(ferror(file)) {
 			err = EIO;
 		}
@@ -223,7 +223,8 @@ struct group *getgrgid(gid_t gid) {
 	});
 
 	if (err) {
-		errno = err;
+		if (err != ENOENT)
+			errno = err;
 		return nullptr;
 	}
 
@@ -237,6 +238,9 @@ int getgrgid_r(gid_t gid, struct group *grp, char *buffer, size_t size, struct g
 	});
 
 	if (err) {
+		if (err == ENOENT)
+			return 0;
+
 		return err;
 	}
 
@@ -257,7 +261,8 @@ struct group *getgrnam(const char *name) {
 	});
 
 	if (err) {
-		errno = err;
+		if (err != ENOENT)
+			errno = err;
 		return nullptr;
 	}
 
@@ -272,6 +277,9 @@ int getgrnam_r(const char *name, struct group *grp, char *buffer, size_t size, s
 	});
 
 	if (err) {
+		if (err == ENOENT)
+			return 0;
+
 		return err;
 	}
 
@@ -386,7 +394,7 @@ int getgrouplist(const char *user, gid_t gid, gid_t *groups, int *ngroups) {
 		return false;
 	});
 
-	if(err != ESRCH) {
+	if(err != ENOENT) {
 		errno = err;
 		return -1;
 	}

diff --git i/options/posix/generic/pwd.cpp w/options/posix/generic/pwd.cpp
index 25a1c8bb..2c4cb68d 100644
--- i/options/posix/generic/pwd.cpp
+++ w/options/posix/generic/pwd.cpp
@@ -158,13 +158,14 @@ struct passwd *getpwnam(const char *name) {
 		}
 	}
 
-	int err = errno;
+	int err = 0;
 	if(ferror(file)) {
 		err = EIO;
 	}
 
 	fclose(file);
-	errno = err;
+	if(err)
+		errno = err;
 	return nullptr;
 }
 
@@ -219,13 +220,14 @@ struct passwd *getpwuid(uid_t uid) {
 		}
 	}
 
-	int err = ESRCH;
+	int err = 0;
 	if(ferror(file)) {
 		err = EIO;
 	}
 
 	fclose(file);
-	errno = err;
+	if(err)
+		errno = err;
 	return nullptr;
 }
 
