diff --git i/options/internal/generic/allocator.cpp w/options/internal/generic/allocator.cpp
index 84bf0e4a..2a36c1d9 100644
--- i/options/internal/generic/allocator.cpp
+++ w/options/internal/generic/allocator.cpp
@@ -27,8 +27,10 @@ MemoryAllocator &getAllocator() {
 // --------------------------------------------------------
 
 uintptr_t VirtualAllocator::map(size_t length) {
-	void *ptr;
-	__ensure(!mlibc::sysdep<AnonAllocate>(length, &ptr));
+	void *ptr = nullptr;
+	if(mlibc::sysdep<AnonAllocate>(length, &ptr))
+		return 0;
+
 	return (uintptr_t)ptr;
 }
 
